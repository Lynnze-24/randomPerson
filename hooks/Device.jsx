import { Platform} from "react-native";

export function useDevice() {
  const { OS: platform } = Platform;

 

  const isWeb = platform === "web";
  const isAndroid = platform === "android";
  const isIOS = platform === "ios";

  return {
    isWeb,
    isAndroid,
    isIOS,
    
  };
}