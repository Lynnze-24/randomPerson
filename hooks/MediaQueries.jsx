import React,{ useMemo } from 'react';
import { useWindowDimensions } from 'react-native'

const useMediaQueries = () => {
    const {  width } = useWindowDimensions();
   
    const isLessThanXS = useMemo(() => {
        return width <= 465;
      }, [width]);
    
      const isLessThanSM = useMemo(() => {
        return width <= 550;
      }, [width]);

      const isLessThanXSM = useMemo(() => {
        return width <= 600;
      }, [width]);
    
      const isLessThanMD = useMemo(() => {
        return width <= 700;
      }, [width]);
    
    
      const isLessThanLG = useMemo(() => {
        return width <= 840;
      }, [width]);

      const isLessThanXLG = useMemo(() => {
        return width <= 1000;
      }, [width]);

      const isLessThanXL = useMemo(() => {
        return width <= 1260;
      }, [width]);

      
    

      return {
        width,
        isLessThanXS,
        isLessThanSM,
        isLessThanXSM,
        isLessThanMD,
        isLessThanLG,
        isLessThanXLG,
        isLessThanXL,
      };
}

export default useMediaQueries;
