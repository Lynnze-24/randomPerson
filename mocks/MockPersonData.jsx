export const mockPersonData = [
    {
        name:{
            title:"Mr",
            first:"Aung",
            last:"Linn"
        },
        picture:{
            thumbnail:1
        },
        email:'bbb@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:122
        }
        
    },
    {
        name:{
            title:"Mrs",
            first:"Htet",
            last:"Htet"
        },
        picture:{
            thumbnail:1
        },
        email:'aaa@gmail.com',
        cell:"999",
        location:{
            city:"Bago",
            country:"Myanmar"
        },
        login:{
            md5:333
        }
        
    }
]