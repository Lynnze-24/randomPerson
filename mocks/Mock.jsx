import { mockPersonData } from "./MockPersonData"

export const mockVal = {personData:[],
    isLoading:false,
    isError:false,
    refreshPerson:jest.fn(),
    filter:'',
    setFilter:jest.fn(),
    refresh:false,
    isTableView:true,
    switchView:jest.fn()
}




export const mockPersonVal = {...mockVal,
    personData: mockPersonData

}