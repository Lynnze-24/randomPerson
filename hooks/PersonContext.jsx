import React, { createContext, useContext, useState,useEffect} from 'react';
import { mockPersonData } from '../mocks/MockPersonData';



const PersonContext = createContext();


export function usePersonContext(){
    return useContext(PersonContext)
}

export function PersonProvider({children,mock,mockFetch}) {
    const initialData = mockFetch?mockPersonData:[]
    const initialLoading = mockFetch?false:true
   
   const [personData, setPersonData] = useState(initialData); // for person data
    
    const[isLoading,setIsLoading] = useState(initialLoading); // for loading state
    const [isError,setIsError] = useState(false);   // handle fetch error 
    const[refresh,setRefresh] = useState(false);    // for refresh button
    const[filter,setFilter] = useState('');         // setting filter word to filter person data
    
    const[isTableView,setTableView] = useState(true) // switch view

    const switchView = () =>{
        setTableView(!isTableView)
    }

    const fetchPerson = ((mock||mockFetch))?()=>{} 
                            :()=>{
                                fetch('https://randomuser.me/api?results=50')
                                .then(res=> res.json())
                                .then(person=> {
                                        if(isError)setIsError(false)
                                        setPersonData(person.results)
                                        setIsLoading(false)
                                    
                                })
                                .catch(()=>{
                                    setIsLoading(false)
                                    setIsError(true);
                                
                                })
                        }
  

    const refreshPerson = ()=>{
        setIsLoading(true);
        setRefresh(!refresh)
    }
      
    useEffect(() => {
           fetchPerson()
    }, [refresh])
    
    const val = mock? mock : {personData,
                                isLoading,
                                isError,
                                refreshPerson,
                                filter,
                                setFilter,
                                refresh,
                                isTableView,
                                switchView}
    
    return (
        <PersonContext.Provider value={val}>
        {children}
      </PersonContext.Provider>
    );
}
