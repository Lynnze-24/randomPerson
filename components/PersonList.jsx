import React,{useMemo} from 'react'
import {View,Text,FlatList,StyleSheet} from 'react-native'
import useMediaQueries from '../hooks/MediaQueries';
import { usePersonContext } from '../hooks/PersonContext';
import cusTheme from '../theme/Theme';
import TableItem from './cards/TableItem';
import TileItem from './cards/TileItem';


const PersonList = () => {

    
    const {isLessThanXL,isLessThanXS,isLessThanLG,width} = useMediaQueries();
    const {personData,filter,isTableView} = usePersonContext();

    const colNum = useMemo(()=>{
      return (isTableView?1:(width>1260)?3:(width>840)?2:1)
    },[width,isTableView]) // for tile view responsive style





    let filteredPerson = personData.filter(item=>{
       return (`${item.name.title} ${item.name.first} ${item.name.last}`).toLowerCase().includes(filter.toLowerCase())
    }) // filtered with query

    let sortedFilterPerson = filteredPerson.sort((a,b)=> a.email.localeCompare(b.email)); 
      // sorted by email column
   
    const renderTableItem = ({ item}) => <TableItem   
                                        name={`${item.name.title} ${item.name.first} ${item.name.last}`} 
                                        avatar={item.picture.thumbnail}
                                        email={item.email}
                                        phone={item.cell}
                                        location={`${item.location.city}, ${item.location.country}`}
                                         />
     const renderTileItem = ({ item}) => <TileItem   
                                         name={`${item.name.title} ${item.name.first} ${item.name.last}`} 
                                         avatar={item.picture.thumbnail}
                                         email={item.email}
                                         phone={item.cell}
                                         location={`${item.location.city}, ${item.location.country}`}
                                          />                                    
    
    return (
        sortedFilterPerson.length?(
                            <FlatList contentContainerStyle={isTableView ?[styles.table,isLessThanXS && styles.tableXS ]
                                                                :[styles.tiles,isLessThanXL && styles.tilesLG ]}
                            data={sortedFilterPerson}
                            testID="personlist"
                            numColumns={colNum}
                            key={colNum}
                            renderItem={isTableView ?renderTableItem:renderTileItem} 
                            keyExtractor={item => item.login.md5} />)
                            :<View  testID="personlist" style={styles.empty}>
                                <Text>There is no such person</Text>
                            </View>
    )
}

const styles = StyleSheet.create({
    table: {
      marginHorizontal:40,
      borderTopColor:cusTheme.lpr,
      borderTopWidth:2,
      flex:1,
      marginVertical:50
   },
   tableXS:{
    marginHorizontal:10,
    marginVertical:20
   },
   tiles: {
        maxWidth:1260,
        paddingVertical:30,
        alignSelf:'center',
       
    },
    tilesLG:{
        justifyContent:'center'
    },
    empty:{
        flex:1,
        justifyContent:'center',
        alignItems:'center'
    }
})

export default PersonList
