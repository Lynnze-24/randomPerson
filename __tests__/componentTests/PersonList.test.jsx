import {render, fireEvent } from "@testing-library/react-native";
import PersonList from "../../components/PersonList";
import { PersonProvider } from "../../hooks/PersonContext";
import { mockVal,mockPersonVal } from "../../mocks/Mock";
import renderer from 'react-test-renderer';



it('matches snapShot', () => {
    const comp = renderer.create(<PersonProvider mock={mockVal}>
                                    <PersonList/>
                                </PersonProvider>).toJSON();
    expect(comp).toMatchSnapshot();
});


it('should render there is no such person if filtered person empty',()=>{ 
    
    const {getByText}=render(
                        <PersonProvider mock={mockVal}>
                         <PersonList/>
                        </PersonProvider>
                        );
                 getByText("There is no such person")       
})

it('should show table view initially i.e. isTableview is true',()=>{ 
    
    const {getAllByTestId,debug}=render(
                        <PersonProvider mock={mockPersonVal}>
                         <PersonList/>
                        </PersonProvider>
                        );
        expect(getAllByTestId('tableItem').length).toBe(2)          
})

it('should show tile view if isTableview is false',()=>{ 
    const mock2 = {...mockPersonVal,isTableView:false}
    const {getAllByTestId,debug}=render(
                        <PersonProvider mock={mock2}>
                         <PersonList/>
                        </PersonProvider>
                        );
           expect(getAllByTestId('tileItem').length).toBe(2)         
})

it('should show only filtered item if search',()=>{ 
    const mock2 = {...mockPersonVal,filter:'Htet'}
    const {getByText,queryByText,getAllByTestId,debug}=render(
                        <PersonProvider mock={mock2}>
                         <PersonList/>
                        </PersonProvider>
                        );
           getByText("Mrs Htet Htet")
           expect(queryByText("Mr Aung Linn")).toBe(null)
           expect(getAllByTestId('tableItem').length).toBe(1)         
})


it('should sort personData with email address',()=>{ 
   
    const {queryAllByTestId}=render(
                        <PersonProvider mock={mockPersonVal}>
                         <PersonList/>
                        </PersonProvider>
                        );
            expect(queryAllByTestId('personName')[0]).toHaveTextContent("Mrs Htet Htet")
            expect(queryAllByTestId('personName')[1]).toHaveTextContent("Mr Aung Linn") // sorted with email because
                                                                                        // Htet Htet email starts with 'a'
                   
})
