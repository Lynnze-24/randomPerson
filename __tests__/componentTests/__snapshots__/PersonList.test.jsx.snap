// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapShot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="personlist"
>
  <Text>
    There is no such person
  </Text>
</View>
`;
